### PR TITLE
Optimize stock queries

### PR DESCRIPTION
**Pull request by jocel1**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Optimize stock & stock movement queries
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-4918
| How to test?  | Run the catalog>stock / movement pages

#### Important guidelines

* Make sure [your local branch is up to date](https://help.github.com/articles/syncing-a-fork/) before commiting your changes!
* Your code MUST respect [our Coding Standards](http://doc.prestashop.com/display/PS16/Coding+Standards) (for code written in PHP, JavaScript, HTML/CSS/Smarty/Twig, SQL)!
* Your commit name MUST respect our [naming convention](http://doc.prestashop.com/display/PS16/How+to+write+a+commit+message)!

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8804)

### CODE REVIEW

**Comment by jocel1** [comment]
Benchmark in progress, but the old queries are so slow it will takes time :)

**Comment by jocel1** [comment]
Well actually blackfire times out with the original query :o
But it's 1760s vs about 8s with the new implementation on my local machine

**Comment by jocel1** [comment]
mm tests are failing, still WIP :)

File: src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php [hunk 1]
```diff
@@ -380,4 +396,131 @@ protected function bindMaxResultsValue(Statement $statement, QueryParamsCollecti
...
+        $rowCount = (int)$statement->fetchColumn();
+        $statement->closeCursor();
+
+        return $rowCount;
+    }
+
+    /**
+     * @param array $row
+     */
+    protected function addCombinationName(array &$row)
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php]: how about rename it to getCombinationName and remove the need of passing an array by reference but the real required information?

File: src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php [hunk 2]
```diff
@@ -380,4 +396,131 @@ protected function bindMaxResultsValue(Statement $statement, QueryParamsCollecti
...
+        $statement->bindValue('language_id', (int)$this->languageId, \PDO::PARAM_INT);
+        $statement->execute();
+        $row['combination_name'] = $statement->fetchColumn(0);
+        $statement->closeCursor();
+    }
+
+    /**
+     * @param array $row
+     */
+    protected function addProductFeatures(array &$row)
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php]: same here, if it's intended and not qn issue : ignore my previous comment about it ;)

File: src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php (hunk 1 above)
Comment (jocel1) [reply src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php]: I've completely changes my approach to fix the remaining issue, so there's no more addCombinationName with the latest revision

**Comment by jocel1** [comment]
Ready to be reviewed!

**Comment by eternoendless** [comment]
Reviewed 2 of 6 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php, line 532 at r3](https://reviewable.io:443/reviews/prestashop/prestashop/8804#-L6bTw6c7TWBbIHSlGc1:-L6bTw6dHaeIknt8uxEq:blmnrir) ([raw file](https://github.com/prestashop/prestashop/blob/4091aec541180df31f9b4792886d9c1c3ee1cb45/src/PrestaShopBundle/Entity/Repository/StockManagementRepository.php#L532)):*
> ```PHP
>     protected function addCombinationsAndFeatures(array $rows)
>     {
>         array_walk($rows, function (&$row) {
> ```

You should use a foreach with a reference instead of array_walk, it can be up to 3x faster!

Test:
https://3v4l.org/pgrNI

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8804)*

**Comment by eternoendless** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8804#-:-L6vL9Gl4aqQxvOF7OZO:bnfp4nl)*

**Review by eternoendless — DISMISSED** [review]

**Review by mickaelandrieu — DISMISSED** [review]

File: tests/phpunit-admin.xml [hunk 3]
```diff
@@ -5,6 +5,8 @@
     <php>
         <env name="SYMFONY_ENV" value="test"/>
         <env name="kernel.environment" value="test" />
+        <const name="_PS_IN_TEST_" value="1"/>
+        <server name="KERNEL_DIR" value="app" />
```
Comment (mickaelandrieu) [inline tests/phpunit-admin.xml]: normalement,

```diff
- <server name="KERNEL_DIR" value="app" />
+ <server name="KERNEL_CLASS" value="AppKernel" />
```

à moins que tu ré-utilises la valeur quelque part mon ami?

Comment (jocel1) [reply tests/phpunit-admin.xml]: non en fait je l'ai juste remis comme avant pour voir si le test qui fail était lié à ce changement, mais en fait non. Donc il doit y avoir autre chose sur develop qui fait fail les tests

**Comment by alegout** [comment]
Reviewed 3 of 3 files at r4, 1 of 1 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8804)*

**Comment by alegout** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8804#-:-L8krQkuCQYJ4m1Xg3a4:bnfp4nl)*

**Review by alegout — DISMISSED** [review]

**Comment by mickaelandrieu** [comment]
@jocel1 one more rebase please, I'll re-add an approval don't worry!

**Comment by marionf** [comment]
@jocel1 
Sorry, but I can't test your PR right now, because there is a bug on the stock page on develop branch

capture du 2018-03-29 14-08-53

**Comment by alegout** [comment]
@marionf normally this is fixed by https://github.com/PrestaShop/PrestaShop/pull/8863 which has been merged few hours ago.

**Comment by marionf** [comment]
So, we need a rebase @jocel1 

rebase

**Comment by jocel1** [comment]
Done!

**Review by mickaelandrieu — DISMISSED** [review]

**Review by mickaelandrieu — APPROVED** [review]

**Comment by jocel1** [comment]
Replace by #8899 (same PR, because of a strange github bug, this PR has some files from develop  not uptodate)

**Pull request closed without merging**